### ISSUE TITLE
Schedulers step before optimizers

### ISSUE THREAD

**Issue by sholalkere** [PyTorchLightning/pytorch-lightning#101]
According to [pytorch docs](https://pytorch.org/docs/stable/optim.html), 

> Learning rate scheduling should be applied after optimizer’s update

Which is also apparent in the warning 
```
UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule.See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
```

And since pytorch>1.1.0 is required, we should either call the scheduler step after `__run_tng_batch()` is called in `__train()` or in `__run_tng_batch()` itself

**Comment by williamFalcon** [comment]
@sidhanthholalkere feel free to make the change, but it has to be at the end of __train, otherwise the steps will be for batch instead of the intended epoch (ie: each step means an epoch, not a batch for an LR_scheduler)

**Comment by MaveriQ** [comment]
Hi

I am getting the same warning with PyTorch Lightning v 1.1.3 when I use OneCycleLR scheduler, passing the interval as 'step'. And I am not sure, but maybe this is why I am getting a very odd behavior from OneCycleLR scheduler. Specifically, changing `final_div_factor` has absolutely no effect on the schedule. 
![Screenshot from 2021-01-09 13-54-59](https://user-images.githubusercontent.com/1519077/104092139-5037ea80-5282-11eb-9928-0638b0ecb114.png)

**Comment by kielnino** [comment]
Same here with PL 1.1.7 and:
```python
scheduler_dict = {
            'scheduler': OneCycleLR(optimizer, max_lr=1e-4, total_steps=int(25e3), verbose=False),
            'interval': 'step',
            'name': 'OneCycleLR'
        }
```
in combination with 
```python
lr_monitor = LearningRateMonitor(logging_interval='step')
```
which gives:

```UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. ```

**Comment by akihironitta** [comment]
It seems this problem still persists as reported in #5587.

**Comment by tmbdev** [comment]
I'm getting the same warning in PL 1.3.3:

```
UserWarning: Seems like `optimizer.step()` has been overridden after learning rate scheduler initialization. Please, make sure to call `optimizer.step()` before `lr_scheduler.step()`. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
```

I use the standard `configure_optimizer` method to return the optimizer and the scheduler.

**Comment by akihironitta** [comment]
@tmbdev @kielnino @MaveriQ Hi, thank you all for your reports. Could anyone provide a full script to reproduce it using [BoringModel](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report_model.py)? I cannot reproduce this behaviour with 1.3.3 or master using my below script.

<details>
<summary>script</summary>

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        # return torch.optim.SGD(self.layer.parameters(), lr=0.1)
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        scheduler = torch.optim.lr_scheduler.OneCycleLR(
            optimizer,
            max_lr=1e-4,
            total_steps=int(25e3),
            verbose=False,
        )
        lr_dict = {
            "scheduler": scheduler,
            "interval": "step",
            "frequency": 1,
            "name": "OneCycleLR",
        }
        return {
            "optimizer": optimizer,
            "lr_scheduler": lr_dict,
        }


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        max_epochs=1,
        weights_summary=None,
    )
    trainer.fit(model, train_dataloader=train_data, val_dataloaders=val_data)


if __name__ == '__main__':
    run()
```

</details>

<details>
<summary>env</summary>

```console
$ wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
$ python collect_env_details.py
* CUDA:
	- GPU:
	- available:         False
	- version:           10.2
* Packages:
	- numpy:             1.20.3
	- pyTorch_debug:     False
	- pyTorch_version:   1.8.1+cu102
	- pytorch-lightning: 1.3.3
	- tqdm:              4.61.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         
	- python:            3.9.4
	- version:           #1 SMP Debian 4.19.181-1 (2021-03-19)
```

**Comment by kielnino** [comment]
Hi @akihironitta, thanks for looking into this. I'm not able to reproduce it with my code either, so I consider this fixed.

**Comment by sagadre** [comment]
If I modify the trainer slightly in @akihironitta's script, setting the gpu affinity to > 1 gpu, I can repro the issue:

`
trainer = Trainer(default_root_dir=os.getcwd(), max_epochs=1, weights_summary=None, gpus=[1, 2])
`

Something to do with ddp?

**Comment by YunchaoLiu** [comment]
Hello, I'm still seeing this problem with PL v1.3.0. any suggestions?

**Comment by BttMA** [comment]
Agree with @akihironitta
It is really a misbehaviour coming from `Trainer(..., precision=16, ...)` (on GPU). I tested it with `pip install pytorch-lightning` and `pip install pytorch-lightning -U` , I got the same warning ``UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.`` when I comment `precision=16` the warning just disappear :)

**Comment by akihironitta** [comment]
Please see #5558 for the behaviour with `precision=16`.

**Comment by BttMA** [comment]
Hello @akihironitta :) 
I dont see any solution in #5558. Could you be more kind and point to us where the solution is or you still working on fixing it and we shall ignore the warning until any further notice ?

**Comment by akihironitta** [comment]
@BttMA Hi, I found the source of the problem of `precision=16` (almost always (maybe?)) causing the warning, but the fix is still wip. I'll add some comments in the linked issue #5558 now.
